bookkeeper-server/src/test/java/org/apache/bookkeeper/util/SubTreeCacheTest.java

```diff
@@ -0,0 +1,321 @@
+            this.guard = guard;
+        }
+
+        @Override
+        public void close() throws Exception {
+            guard.close();
+        }
+    }
+
+    TestWatchGuard setWatchWithGuard() {
+        TestWatchGuard watch = new TestWatchGuard();
+        watch.setGuard(cache.registerWatcherWithGuard(watch));
+        return watch;
+    }
+
+    void readAssertChildren(String path, String[] children) throws KeeperException, InterruptedException {
+        SortedSet<String> shouldBe = new TreeSet<String>(Arrays.asList(children));
+        List<String> returned = cache.getChildren(path);
+        SortedSet<String> is = new TreeSet<String>(returned);
+        returned.clear(); // trip up implementations which return an internal reference
+        Assert.assertEquals(shouldBe, is);
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        String[] preCreate =
+                {"/a"
+                        , "/a/a"
+                        , "/a/a/a"
+                        , "/a/a/b"
+                        , "/a/b"
+                        , "/a/c"
+                        , "/b"
+                        , "/b/a"
+                };
+        for (String path : preCreate) {
+            tree.createNode(path);
+        }
+    }
+
+    @Test(timeout=5000)
+    public void testNoUpdate() throws Exception {
+        TestWatch watch = setWatch();
+        readAssertChildren("/a/a", new String[]{"a", "b"});
+        assertNotFired(watch);
+    }
+
+    @Test(timeout=5000)
+    public void testSingleCreate() throws Exception {
+        TestWatch watch = setWatch();
+        readAssertChildren("/a/a", new String[]{"a", "b"});
+        tree.createNode("/a/a/c");
+        assertFired(watch);
+    }
+
+    @Test(timeout=5000)
+    public void testSingleRemoval() throws Exception {
+        TestWatch watch = setWatch();
+        readAssertChildren("/a/a", new String[]{"a", "b"});
+        tree.removeNode("/a/a/b");
+        assertFired(watch);
+    }
+
+    @Test(timeout=5000)
+    public void testCancelation() throws Exception {
+        TestWatch watch = setWatch();
+        readAssertChildren("/a/a", new String[]{"a", "b"});
+        cache.cancelWatcher(watch);
+        tree.createNode("/a/a/c");
+        assertNotFired(watch);
+    }
+
+    @Test(timeout=5000)
+    public void testGuardCancelation() throws Exception {
+        TestWatch watch;
+        try (TestWatchGuard guard = setWatchWithGuard()) {
+            readAssertChildren("/a/a", new String[]{"a", "b"});
+            watch = guard;
+        }
+        tree.createNode("/a/a/c");
+        assertNotFired(watch);
+    }
+
+    @Test(timeout=5000)
+    public void testGuardCancelationExceptional() throws Exception {
+        TestWatch watch = null;
+        try (TestWatchGuard guard = setWatchWithGuard()) {
+            watch = guard;
+            readAssertChildren("/z/a", new String[]{});
+        } catch (Exception e) {
```

Comment:
No, the exception is meant to fire and the test is that the watch is canceled when it does (hence the assertNotFired below).